app/controllers/image_controller.rb

```diff
@@ -364,62 +410,96 @@ def process_image(args, upload)
   # Outputs: @image, @licenses
   def edit_image
-    pass_query_params
-    return unless (@image = find_or_goto_index(Image, params[:id].to_s))
+    return unless image_exists
 
-    @licenses = License.current_names_and_ids(@image.license)
-    if !check_permission!(@image)
-      redirect_with_query(action: "show_image", id: @image)
-    elsif request.method != "POST"
-      init_project_vars_for_add_or_edit(@image)
+    set_licenses_ivar
+    check_image_permission
+    if request.method == "POST"
+      update_image
     else
-      @image.attributes = whitelisted_image_params
-      xargs = {}
-      xargs[:set_date] = @image.when if @image.when_changed?
-      xargs[:set_notes] = @image.notes if @image.notes_changed?
-      if @image.copyright_holder_changed?
-        xargs[:set_copyright_holder] = @image.copyright_holder
-      end
-      if @image.original_name_changed?
-        xargs[:set_original_name] = @image.original_name
-      end
-      xargs[:set_license] = @image.license if @image.license_id_changed?
-      done = false
-      if xargs.empty?
-        if update_projects(@image, params[:project])
-          flash_notice(:runtime_image_edit_success.t(id: @image.id))
-        else
-          flash_notice(:runtime_no_changes.t)
-        end
-        done = true
-      elsif !@image.save
-        flash_object_errors(@image)
-      else
-        xargs[:id] = @image
-        @image.log_update
-        flash_notice(:runtime_image_edit_success.t(id: @image.id))
-        update_projects(@image, params[:project])
-        done = true
-      end
-      if done
-        redirect_with_query(action: "show_image", id: @image.id)
-      else
-        init_project_vars_for_reload(@image)
-      end
+      init_project_vars_for_add_or_edit(@image)
     end
   end
 
   ##############################################################################
 
   private
 
+  def update_image
+    return unless image_exists
+
+    set_licenses_ivar
+    check_image_permission
+
+    @image.attributes = whitelisted_image_params
+    xargs = set_xargs
+
+    done = figure_out_if_done(xargs)
+    if done
+      redirect_with_query(action: "show_image", id: @image.id)
+    else
+      init_project_vars_for_reload(@image)
+    end
+  end
+
+  def set_licenses_ivar
+    @licenses = License.current_names_and_ids(@image.license)
+  end
+
+  def check_image_permission
+    redirect_with_query(action: "show_image", id: @image) unless
+      check_permission!(@image)
+  end
+
+  def set_xargs
+    xargs = {}
+    xargs[:set_date] = @image.when if @image.when_changed?
+    xargs[:set_notes] = @image.notes if @image.notes_changed?
+    if @image.copyright_holder_changed?
+      xargs[:set_copyright_holder] = @image.copyright_holder
+    end
+    if @image.original_name_changed?
+      xargs[:set_original_name] = @image.original_name
+    end
+    xargs[:set_license] = @image.license if @image.license_id_changed?
+    xargs
+  end
+
+  def figure_out_if_done(xargs = {})
```

Comment:
This could be done much more simply without xargs by just checking if any important columns changed: when_changed?, notes_changed?, copyright_holder_changed?, original_name_changed?, license_id_changed?